File: QuickSorts/OCaml/functionQuickSort.ml
(*
 * Geneva Smith 
 * ----------------------------------------------------------------------------
 * This program sorts a list of anything using the quicksort algorithm, a
 * provided comparision function, and a provided partition function that uses
 * the comparision function.
 * Tested types: int, float, char, string
 *)

(*
 * Function: print_char_list
 * Arguments: list of chars
 * ----------------------------------------------------------------------------
 * This function takes in a list of chars and pattern matches it with one of 
 * two cases:
 *  Case 1: The list is empty so print an empty string
 *  Case 2: The list is non-empty, so print the  head char and recursively call
 *          this function on the tail.
 *)
let rec print_char_list cList =
	match cList with
	| [] -> print_string ""
	| head :: tail -> print_char head; print_string " "; print_char_list tail
	;;
	
(*
 * Function: functionQuickSort
 * Arguments: list of anything, comparison function, partition function
 * ----------------------------------------------------------------------------
 * This function takes in a comparison function and list of anything which is 
 * pattern matched with one of two cases:
 *  Case 1: The list is empty, return empty list
 *  Case 2: The list is non-empty, take the head of the list as the pivot and
 *          filter the rest into two more lists that are less than the pivot
 *          (leftSide) and greater/equal to the pivot (rightSide) using the 
 *          provided partition and comparision functions. It then uses the
 *          two new lists to recursively call this function and append it to
 *          the pivot.
 *)
let rec functionQuickSort intList comp part =
	match intList with
	| [] -> []
	| pivot :: tail -> 
		let leftSide, rightSide = (part pivot comp) tail in
		(functionQuickSort leftSide comp part) @ [pivot] @ (functionQuickSort rightSide comp part)
	;;


(* Test Cases *)
let test1 = [ ];;
let test2 = [ 1; ];;
let test3 = [ 1; 2; 3; ];;
let test4 = [ 3; 2; 1; ];;
let test5 = [ 3; 0; 1; 8; 7; 2; 5; 4; 9; 6; ];;
let test6 = [ 3.0; 0.0; 1.0; 8.0; 7.0; 2.0; 5.0; 4.0; 9.0; 6.0; ];;
let test7 = [ 4.0; 2.0; 7.0; 3.0; 7.0; 1.0; ];;
let test8 = [ 'd'; 'a'; 'b'; 'i'; 'h'; 'c'; 'f'; 'e'; 'j'; 'g'; ];;
let test9 = [ "cat"; "bat"; "at"; "gnat"; ];;

(* Comparision functions *)
let ascend = fun x y ->	x <= y ;;
let descend = fun x y -> x >= y ;;

(* Partition function *)
let part = fun pivot comp -> List.partition (fun x -> comp x pivot) ;;

let test1Result = functionQuickSort test1  ascend part ;;
let test2Result = functionQuickSort test2 descend part ;;
let test3Result = functionQuickSort test3  ascend part ;;
let test4Result = functionQuickSort test4 descend part ;;
let test5Result = functionQuickSort test5  ascend part ;;
let test6Result = functionQuickSort test6 descend part ;;
let test7Result = functionQuickSort test7  ascend part ;;
let test8Result = functionQuickSort test8 descend part ;;
let test9Result = functionQuickSort test9  ascend part ;;

print_endline("\nInteger Lists\n");;

print_string("Test 1 Ascend before: [" ^ String.concat " " (List.map string_of_int test1)); print_endline("]");;
print_string("Test 1 Ascend after: [" ^ String.concat " " (List.map string_of_int test1Result)); print_endline("]\n");;

print_string("Test 2 Descend before: [" ^ String.concat " " (List.map string_of_int test2)); print_endline("]");;
print_string("Test 2 Descend after: [" ^ String.concat " " (List.map string_of_int test2Result)); print_endline("]\n");;

print_string("Test 3 Ascend before: [" ^ String.concat " " (List.map string_of_int test3)); print_endline("]");;
print_string("Test 3 Ascend after: [" ^ String.concat " " (List.map string_of_int test3Result)); print_endline("]\n");;

print_string("Test 4 Descend before: [" ^ String.concat " " (List.map string_of_int test4)); print_endline("]");;
print_string("Test 4 Descend after: [" ^ String.concat " " (List.map string_of_int test4Result)); print_endline("]\n");;

print_string("Test 5 Ascend before: [" ^ String.concat " " (List.map string_of_int test5)); print_endline("]");;
print_string("Test 5 Ascend after: [" ^ String.concat " " (List.map string_of_int test5Result)); print_endline("]\n");;

print_endline("\nFloat Lists\n");;

print_string("Test 6 Descend before: [" ^ String.concat " " (List.map string_of_float test6)); print_endline("]");;
print_string("Test 6 Descend after: [" ^ String.concat " " (List.map string_of_float test6Result)); print_endline("]\n");;

print_string("Test 7 Ascend before: [" ^ String.concat " " (List.map string_of_float test7)); print_endline("]");;
print_string("Test 7 Ascend after: [" ^ String.concat " " (List.map string_of_float test7Result)); print_endline("]\n");;

print_endline("\nChar Lists\n");;

print_string("Test 8 Descend before: [ "); print_char_list test8; print_endline("]");
print_string("Test 8 Descend after: [ "); print_char_list test8Result; print_endline("]\n");;

print_endline("\nString Lists\n");;

print_string("Test 9 Ascend before: [ " ^ String.concat " " test9); print_endline("]");
print_string("Test 9 Ascend after: [ " ^ String.concat " " test9Result); print_endline("]\n");;

File: QuickSorts/Ruby/FunctionQuickSort.rb
# Geneva Smith
# ----------------------------------------------------------------------------
# This program sorts a list of anything using the quicksort algorithm, a
# provided comparision function, and a provided partition function. If
# functions are not provided, defaults are used.
# Tested types: int, float, char, string

class QuickSort
	# Public function: FunctionQuickSort
	# ----------------------------------------------------------------------------
	# A boilerplate for the quicksort algorithm that sorts a list of 
	# anything and uses a provided partition and comparision function.
	def FuntionQuickSort (array=[], compare=Proc.new { |a, b| a <= b }, partition=nil)
		if array.length < 2
			return array
		else
			array = FuntionQS(array, 0, array.length - 1, compare, partition)
		end
		
		return array
	end

	# Privatefunction: FunctionQS
	# ----------------------------------------------------------------------------
	# The  quicksort algorithm.
	private
	def FuntionQS (array=[], first=0, last=0, compare, partition)
		if first < last
			if partition == nil
				pivot = FuntionPartition(array, first, last, compare)
			else
				pivot = partition.call(array, first, last, compare)
			end
			
			FuntionQS(array, first, pivot - 1, compare, partition)
			FuntionQS(array, pivot + 1, last, compare, partition)
		end
		
		return array
	end

	# Privatefunction: TemplatePartition
	# ----------------------------------------------------------------------------
	# This function is used if a paritition function is not provided.
	# Divides an array of integers into two groups based on the 
	# provided comparision function.
	# <left> and <right> track indexes in the array to compare with the <pivot>
	private
	def FuntionPartition (array=[], first=0, last=0, compare)
		left = first
		pivot = first
		right = last
		
		while left != right
			while compare.call(array[pivot], array[right]) && pivot != right
				right = right -1
			end
			
			if pivot != right
				array[pivot], array[right] = array[right], array[pivot]
				pivot = right
			end
			
			while compare.call(array[left], array[pivot]) && pivot != left
				left = left + 1
			end
			
			if left != pivot
				array[left], array[pivot] = array[pivot], array[left]
				pivot = left
			end
		end
		
		return pivot
	end
end

# MAIN (Test Cases)
if __FILE__ == $0
	QS = QuickSort.new()
	test1 = [];
	test2 = [1];
	test3 = [1, 2, 3];
	test4 = [3, 2, 1];
	test5 = [3, 0, 1, 8, 7, 2, 5, 4, 9, 6];
	test6 = [3.0, 0.0, 1.0, 8.0, 7.0, 2.0, 5.0, 4.0, 9.0, 6.0];
	test7 = [4.0, 2.0, 7.0, 3.0, 7.0, 1.0];
	test8 = ['d', 'a', 'b', 'i', 'h', 'c', 'f', 'e', 'j', 'g'];
	test9 = [ "cat", "bat", "at", "gnat" ];
	
	ascending = Proc.new { |a, b| a <= b }
	descending = Proc.new { |a, b| a >= b }
	customPartition = Proc.new { | array, first, last, compare |
		left = first
		pivot = first
		right = last
		
		while left < right
			while compare.call(array[pivot], array[right]) && pivot != right
				right = right -1
			end
			
			while compare.call(array[left], array[pivot]) && pivot != left
				left = left + 1
			end
			
			if left < right
				array[left], array[right] = array[right], array[left]
			end
		end
		
		right = right	#Forces Proc to return the value of <right>
	}

	puts "START"
	puts ""
	puts "Integer QuickSorts"
	puts ""
	
	a = test1.dup
	puts "Test 1 Ascending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, ascending)
	puts "Test 1 Ascending, Default After: #{a.inspect}"
	puts ""
	
	b = test1.dup
	puts "Test 1 Ascending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, ascending, customPartition)
	puts "Test 1 Ascending, Custom After: #{b.inspect}"
	puts ""
	
	a = test2.dup
	puts "Test 2 Descending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, descending)
	puts "Test 2 Descending, Default After: #{a.inspect}"
	puts ""
	
	b = test2.dup
	puts "Test 2 Descending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, descending, customPartition)
	puts "Test 2 Descending, Custom After: #{b.inspect}"
	puts ""
	
	a = test3.dup
	puts "Test 3 Ascending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, ascending)
	puts "Test 3 Ascending, Default After: #{a.inspect}"
	puts ""
	
	b = test3.dup
	puts "Test 3 Ascending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, ascending, customPartition)
	puts "Test 3 Ascending, Custom After: #{b.inspect}"
	puts ""
	
	a = test4.dup
	puts "Test 4 Descending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, descending)
	puts "Test 4 Descending, Default After: #{a.inspect}"
	puts ""
	
	b = test4.dup
	puts "Test 4 Descending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, descending, customPartition)
	puts "Test 4 Descending, Custom After: #{b.inspect}"
	puts ""
	
	a = test5.dup
	puts "Test 5 Ascending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, ascending)
	puts "Test 5 Ascending, Default After: #{a.inspect}"
	puts ""
	
	b = test5.dup
	puts "Test 5 Ascending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, ascending, customPartition)
	puts "Test 5 Ascending, Custom After: #{b.inspect}"
	puts ""
	puts ""
	puts "Double/Float QuickSorts"
	puts""
	
	a = test6.dup
	puts "Test 6 Descending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, descending)
	puts "Test 6 Descending, Default After: #{a.inspect}"
	puts ""
	
	b = test6.dup
	puts "Test 6 Descending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, descending, customPartition)
	puts "Test 6 Descending, Custom After: #{b.inspect}"
	puts ""
	
	a = test7.dup
	puts "Test 7 Ascending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, ascending)
	puts "Test 7 Ascending, Default After: #{a.inspect}"
	puts ""
	
	b = test7.dup
	puts "Test 7 Ascending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, ascending, customPartition)
	puts "Test 7 Ascending, Custom After: #{b.inspect}"
	puts ""
	puts ""
	puts "Character QuickSorts"
	puts ""
	
	a = test8.dup
	puts "Test 8 Descending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, descending)
	puts "Test 8 Descending, Default After: #{a.inspect}"
	puts ""
	
	b = test8.dup
	puts "Test 8 Descending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, descending, customPartition)
	puts "Test 8 Descending, Custom After: #{b.inspect}"
	puts ""
	puts ""
	puts "String QuickSorts"
	puts ""
	
	a = test9.dup
	puts "Test 9 Descending, Default Before: #{a.inspect}"
	QS.FuntionQuickSort(a, descending)
	puts "Test 9 Descending, Default After: #{a.inspect}"
	puts ""
	
	b = test9.dup
	puts "Test 9 Descending, Custom Before: #{b.inspect}"
	QS.FuntionQuickSort(b, descending, customPartition)
	puts "Test 9 Descending, Custom After: #{b.inspect}"
	puts ""
	puts "END"
end